File: 528.random-pick-with-weight/solution.kt
class Solution(w: IntArray) {

    private val prefixSums = mutableListOf<Int>()
    private var totalSum = 0

    init {
        for (weight in w) {
            totalSum += weight
            prefixSums.add(totalSum)
        }
    }

    fun pickIndex(): Int {
        val target = (1..totalSum).random()

        // binarySearch returns the index of the element if found.
        // If not found, it returns (-(insertion point) - 1).
        val index = prefixSums.binarySearch(target)
        
        return if (index >= 0) index else -(index + 1)
    }

}

/**
 * Your Solution object will be instantiated and called as such:
 * var obj = Solution(w)
 * var param_1 = obj.pickIndex()
 */

File: 528.random-pick-with-weight/solution.go
type Solution struct {
    prefix []int
    limit  int
}


func Constructor(w []int) Solution {
    prefix := make([]int, 0, len(w))
    sum := 0
    for _, weight := range w {
        sum += weight
        prefix = append(prefix, sum)
    }
    return Solution{
        prefix: prefix,
        limit:  sum,
    }
}


func (this *Solution) PickIndex() int {
    r := rand.Intn(this.limit) + 1
    i := sort.Search(len(this.prefix), func(i int) bool { return this.prefix[i] >= r })
    return i
}


/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(w);
 * param_1 := obj.PickIndex();
 */